### DISCUSSION TITLE
Timebase-client library error when executing "OVER TIME"

### DISCUSSION THREAD

**Discussion by ShivaniSK** [finos/TimeBase-CE#255] [Q&A] [answered]
Hi @alex-karpovich ,

**NOTE**: All the below queries work fine in the Timebase shell. Issue occurs only when executing using the code with timebase-client library. We are using timebase-client 6.1.17 version which is the latest version available in maven. Also we are using version 6.1.19 for Timebase server.

```
// https://mvnrepository.com/artifact/org.finos.timebase-ce/timebase-client
implementation group: 'org.finos.timebase-ce', name: 'timebase-client', version: '6.1.17'
```

The market_volume stream data is as below:

`select * from market_volume where timestamp BETWEEN 1661471763000 AND 1661471999999 AND SYMBOL == 'EURUSD'`

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kanashia/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kanashia/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

symbol | timestamp | MarketVolumeMessage.trade_type | MarketVolumeMessage.venue | MarketVolumeMessage.volume
-- | -- | -- | -- | --
EURUSD | 2022-08-25T23:57:00.000Z | FxSpot | Market Colour | 6.75
EURUSD | 2022-08-25T23:58:00.000Z | FxSpot | Market Colour | 26.63
EURUSD | 2022-08-25T23:58:00.000Z | FxSpot | EBS | 1
EURUSD | 2022-08-25T23:59:00.000Z | FxSpot | Market Colour | 1.38
EURUSD | 2022-08-25T23:59:00.000Z | FxSpot | EBS | 9

</body>

</html>

As seen in above data, the **last record** in the market_volume stream for symbol **EURUSD** is for venue **EBS**
We are currently facing issues when executing below query with **venue=EBS where clause** using the timebase-client library. 

`select sum{} (volume) as volume, venue from market_volume **OVER TIME (1m)** where timestamp BETWEEN 1661385600000 AND 1661471999999 AND SYMBOL == 'EURUSD' AND venue == 'EBS' GROUP BY venue `

below is the code snippet used for executing the query

```
db.executeQuery(query, getSelectionOptions(resource, timeSpan)).use { cursor ->
            while (cursor.next()) {
                ..
                ..
            }
        }
```
On executing the above query even after reaching the last record, the timebase cursor stays open. And it gives back the first record again and goes into the loop. We are not getting any response here and the thread keeps running.

On debugging further I can see that even after reaching the last record the "while(cursor.next())" is not returning END_OF_STREAM.

Inside `TickCursorClient.java` file

```
private NextResult                  next (boolean throwable) {

..
..

if (nextStreamMsgState == NextMessageState.NOT_QUERIED) {
                result = queryStream();
				
	       // On reaching the last record, the above queryStream function should return NextResult.END_OF_CURSOR, 
however it returns NEXTRESULT.OK all the times.

}
}
```

**NOTE:** The query works fine with other venues in where clause like below:

`select sum{} (volume) as volume, venue from market_volume OVER TIME (1m) where timestamp BETWEEN 1661385600000 AND 1661471999999 AND SYMBOL == 'ABC' AND venue == 'Market Colour' GROUP BY venue`

It also works fine when we remove the "where clause" for "venue" like below (which is our default query which gets executed).

`select sum{} (volume) as volume, venue from market_volume OVER TIME (1m) where timestamp BETWEEN 1661385600000 AND 1661471999999 AND SYMBOL == 'ABC' GROUP BY venue`

**Issue occurs only when adding where clause with the last record as value. In our case "venue=EBS"**

Can you please look into this issue, as its resulting into lot of open cursors in timebase.

Regards,
Shivani

**Comment by alex-karpovich** [comment]
Greetings, @ShivaniSK 
we fixed same issue in Enterprise version recently.

I will push fix soon.

**Reply by ShivaniSK** [reply]
Thanks @alex-karpovich . Do we have any timeline by when the fix will be pushed?

**Reply by alex-karpovich** [reply]
On Monday it will available, will try to do on weekend

**Reply by alex-karpovich** [reply, marked as answer]
Greetings, @ShivaniSK 

Please use https://github.com/finos/TimeBase-CE/releases/tag/6.1.20
